src/jit/hwintrinsicxarch.cpp

```diff
@@ -838,8 +815,37 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
     }
 
     // table-driven importer of simple intrinsics
-    if (isTableDriven)
+    if (impIsTableDrivenHWIntrinsic(intrinsic, category))
     {
+        if ((category == HW_Category_MemoryStore) || HWIntrinsicInfo::BaseTypeFromFirstArg(intrinsic) ||
+            HWIntrinsicInfo::BaseTypeFromSecondArg(intrinsic))
+        {
+            CORINFO_ARG_LIST_HANDLE arg = sig->args;
+
+            if ((category == HW_Category_MemoryStore) || HWIntrinsicInfo::BaseTypeFromSecondArg(intrinsic))
+            {
+                arg = info.compCompHnd->getArgNext(arg);
+            }
+
+            CORINFO_CLASS_HANDLE argClass = info.compCompHnd->getArgClass(sig, arg);
+            baseType                      = getBaseTypeAndSizeOfSIMDType(argClass);
+
+            if (baseType == TYP_UNKNOWN) // the argument is not a vector
+            {
+                CORINFO_CLASS_HANDLE tmpClass;
+                CorInfoType          corInfoType = strip(info.compCompHnd->getArgType(sig, arg, &tmpClass));
+
+                if (corInfoType == CORINFO_TYPE_PTR)
+                {
+                    corInfoType = info.compCompHnd->getChildType(argClass, &tmpClass);
```

Comment:
I apologize for not responding to the questions you posted on the issue. It's never been clear to me why we focus on getting the result type from the arguments, rather than from the return type. Is it just that this code was largely modeled on the `Vector<T>` type, where they usually correspond? I think that the return type could universally be obtained from `sig->retTypeClass`, or is there something I'm missing?

Reply:
> It's never been clear to me why we focus on getting the result type from the arguments, rather than from the return type

We do normally get the base type from `sig->retTypeClass`. However, there are a number of methods where `sig->retTypeClass` is the same for multiple overloads of a single method, in which case you have to differentiate by the arguments. Such methods are generally denoted by the `BaseTypeFromFirstArg` or `BaseTypeFromSecondArg` flag.

For example, we have:
* `Vector256<int> ConvertToVector256Int32(sbyte* address)`
* `Vector256<int> ConvertToVector256Int32(byte* address)`
* `Vector256<int> ConvertToVector256Int32(short* address)`
* `Vector256<int> ConvertToVector256Int32(ushort* address)`

All four of these methods have the same name and same return type, but they each generate a different instruction. So we have to look at the type of the first argument to determine the instructions:
* `VPMOVSXBD`
* `VPMOVZXBD`
* `VPMOVSXWD`
* `VPMOVZXWD`

Reply:
Great - thanks for clarifying!